codex/erasure/asyncbackend.nim

```diff
@@ -0,0 +1,62 @@
+import pkg/taskpools
+
+import pkg/chronos
+import pkg/chronos/threadsync
+
+import pkg/chronicles
+import pkg/questionable
+import pkg/questionable/results
+
+import ./backend
+
+type
+  EncoderArgs* = object
+    backend*: ptr EncoderBackend
+    data*: ptr seq[seq[byte]]
+    parity*: ptr seq[seq[byte]]
```

Comment:
Even though you're passing this via a pointer, that the type is a seq of seq's may eventually cause GC corruption. The reason is that even if you pass outer seq by pointer and copy it, the inner seq's will still be tied to the main thread's GC. Even iterating the outer seq to copy the inner seq's to new ones may end linking the original inner seq's memory to the worker threads heap, causing the corruption.

I'm not 100% certain what will happen in this scenario since it's both refc and a seq of seq's. Though pretty much anytime I've tried passing seq[byte] or strings in refc it'd result in GC corruption. IIRC, @dryajov ran into similar crashes as well. The most reliable I found is to use `ptr UncheckedArray[byte]` and a len.

It may not occur in this instance if `encode` args are passed purely as pointers by the compiler internally and the pointers are used to copy the data. However, that may depend on the compiler flags and version as well as the encode parameter types.

Note, the GC corruption is only triggered on GC collection, but you can add a `GC_fullCollect` to force the GC to run. I did this in a loop and just ran tests several times to trigger the error. I think GC_fullCollect may need to run on the worker thread, maybe the main thread too. I can't recall that.

Reply:
Unfortunately I don't have an easy recommendations. 

The simplest would probably be to flatten it into a single `seq[byte]` and pass that as an `(ptr UncheckedArray[byte], int)` and use iterators to reconstruct the shape in the worker thread.

Another option could be to use something like the `DataBuffer` SharedPtrs I setup in datastore which can be passed safely between threads as they use atomics and sharedMemory.

The last option I see would be to create a sequence:

```nim
type
  ArrayHolder*[T] = object
    data*: ptr UncheckedArray[T]
    size*: int

  EncoderArgs* = object
     backend*: ptr EncoderBackend
     data*: ArrayHolder[ArrayHolder[byte]]

var data: seq[ArrayHolder[byte]] = origData.mapIt(ArrayHolder[byte](data: addr(it[0]), size: it.len))
var dataHolder = ArrayHolder[ArrayHolder[byte](data: addr(data[0]), size: data.len())
```

That seems pretty gross though. Flattening it with existing serialization primitives would be probably be easier.

Reply:
Thanks for the support on the task. For viewers: I discussed this issue with @elcritch and passing `seq[seq[byte]]` to a thread does a deep copy - so it should be safe to simply pass nested seqs (tested on Nim 1.6.16)

Reply:
> Thanks for the support on the task. For viewers: I discussed this issue with @elcritch and passing `seq[seq[byte]]` to a thread does a deep copy - so it should be safe to simply pass nested seqs (tested on Nim 1.6.16)

@tbekas note that Balazs was running into odd corruptions and crashes when passing seq's as arguments to taskpools, similar to this case. So it'll be something to keep an eye on. Moving to ARC/ORC resolved his issues.

Reply:
> Thanks for the support on the task. For viewers: I discussed this issue with @elcritch and passing seq[seq[byte]] to a thread does a deep copy - so it should be safe to simply pass nested seqs (tested on Nim 1.6.16)

I would recommend against this, there is really no reason to do a deep copy at all and passing by pointer should work, there is also no reason to flatten anything, arrays of arrays of arrays (etc..) are already laid out in a flattened manned in memory, so just getting the pointer to the first element and moving it by the number of bytes of each subelement should work just fine.

Reply:
Passing a pointer can lead to memory error as @elcritch showed in this example
https://github.com/codex-storage/apatheia/pull/9/files#diff-764eb5308e6de6c589e7bd145e1863ba9a52b789a75bdef770099e8a3eaf7807

Reply:
> Passing a pointer can lead to memory error as @elcritch showed in this example (and other examples from this PR) https://github.com/codex-storage/apatheia/pull/9/files#diff-764eb5308e6de6c589e7bd145e1863ba9a52b789a75bdef770099e8a3eaf7807

Note that that failure requires the current future to be cancelled, which is done `      await runTest(tp, sig).wait(100.milliseconds)`.